ovs-dpdk.md

```diff
@@ -0,0 +1,170 @@
+Number of pages can vary according to usage requirements. 
+Hugepages must be marked as shared in order to enable host and guest communication, as packets are transmitted and received on buffers allocated on hugepages memory. 
+
+Review custom properties
+```
+# engine-config -g UserDefinedVMProperties
+```
+Add engine custom properties for hugepages support (followed by engine restart)
+```
+# engine-config -s "UserDefinedVMProperties=hugepages=^.*$;hugepages_shared=(true|false)"
+```
+ 
+
+### Enable DPDK in OVS:
+
+ovs-vscwhitchd should get a few parameters in order to enable DPDK functionality.
+
+* Set ```dpdk-init``` so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+
+* ```dpdk-socket-mem```: a comma seprated list of amount of memory to allocate from each NUMA node.
+For example, in a dual socket system, we allocate 1GB from NUMA node 0 and 0GB from the second:
+
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0" 
+```
+
+* ```dpdk-lcore-mask```: Specifies the CPU cores on which dpdk lcore threads should be spawned and expects hex string.
+
+* ```pmd-cpu-mask```: PMD (poll mode driver) threads are used to poll DPDK devices for new packets instead of using using the NIC driver that interacts with the kernel. 
+To better scale the workloads across cores, multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. Ensure that cores from the same NUMA nodes are enabled.
+eg: To spawn 2 pmd threads and pin them to cores 1, 2:
+
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+* Add the following entry to  /etc/sysconfig/openvswitch:
+```
+"DPDK_OPTIONS='-c 110 -n 2 --socket-mem 1024,0”
+```
+The pmd-cpu-mask and -c coremask have to be both specified, with the same value.
+ 
+ Restart openvswitch service
+```
+# sudo systemctl restart openvswitch
+```
+ 
+Verify DPDK is enabled:
+```
+# ovs-vsctl get Open_vSwitch . iface_types
+[dpdk, dpdkr, dpdkvhostuser, dpdkvhostuserclient, geneve, gre, internal, ipsec_gre, lisp, patch, stt, system, tap, vxlan]
+```
+
+## OVS-DPDK in oVirt
+ 
+* Upon creating a new cluster, choose OVS as the switch type.
+* When setting up host networks, DPDK-bound NICs are shown in the dialog,as well as Linux-controlled ones
+* When creating a network on DPDK device, an OVS bridge with datapath_type=’netdev’ will be automatically attached to dpdk device will be created.
+
+[ovs-vsctl]: (/images/ovs-dpdk.png)
+ 
+### VMs
+
+VMs will be created with vhostuser network interface. In order to support vhostuser interfaces, hugepages should be configured on the VM.
+ 
+In oVirt GUI, configure hugepages size, number of pages to use and enable shared hugepages.
```

Comment:
What about using VDSM reserved pages? Are ovs-dpdk hugepages consumed on boot or later on? oVirt could use all allocated pages before ovs-dpdk claims them in that case.